File: Problem Solving/boj/Segment tree/13977-04-05-2023.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

const int MOD = 1'000'000'007;
const int MX = 4'000'000;
const int ROOT = 1;

int n, r, i, j;
ll muls[4 * MX];

ll init(int st, int en, int cur) {
  if(st == en) return muls[cur] = st;
  int mid = (st + en) / 2;
  ll lf = init(st, mid, cur*2);
  ll rf = init(mid + 1, en, cur*2 + 1);
  return muls[cur] = (lf * rf) % MOD;
}

ll mul(int st, int en, int cur) {
  if(en < i || j < st) return 1;
  if(i <= st && en <= j) return muls[cur];
  int mid = (st + en) / 2;
  return mul(st, mid, cur*2) * mul(mid + 1, en, cur*2 + 1) % MOD;
}

ll pow(ll no, int po) { // return no^po
  if(po == 1) return no;
  ll hpow = pow(no, po / 2);
  hpow *= hpow; hpow %= MOD;
  ll ans = 1;
  if(po % 2) ans = no;
  return (ans * hpow) % MOD; 
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  init(0, MX, ROOT);

  int m; cin >> m;
  while(m--) {
    cin >> n >> r;
    i = 1, j = r;
    ll rfact = mul(0, MX, ROOT);
    ll inv_rfact = pow(rfact, MOD - 2);

    i = n - r + 1, j = n;
    ll nPr = mul(0, MX, ROOT);
    ll ans = (nPr * inv_rfact) % MOD;
    cout << ans << '\n';
  }
}

File: problem-solving/boj/segment-tree/13977-04-05-2023.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.math.BigInteger;
import java.util.StringTokenizer;

public class Main {
    private static final BigInteger MOD =  new BigInteger("1000000007");
    private static final int MX = 4_000_000;
    private static final int ROOT = 1;
    private static BigInteger[] muls = new BigInteger[MX * 4];

    static int i, j;

    public static BigInteger init(int st, int en, int cur) {
        if(st == en) return muls[cur] = BigInteger.valueOf(st);
        int mid = (st + en) / 2;
        BigInteger lt = init(st, mid, cur * 2);
        BigInteger rt = init(mid + 1, en, cur * 2 + 1);
        return muls[cur] = lt.multiply(rt);
    }

    public static BigInteger mul(int st, int en, int cur) {
        if(en < i || st > j) return BigInteger.valueOf(1);
        if(i <= st && en <= j) return muls[cur];

        int mid = (st + en) / 2;
        BigInteger lt = mul(st, mid, cur * 2);
        BigInteger rt = mul(mid + 1, en, cur * 2 + 1);
        return lt.multiply(rt);
    }

    public static void main(String args[]) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        init(0, MX, ROOT);

        StringTokenizer st = new StringTokenizer(br.readLine());
        int m = Integer.parseInt(st.nextToken());
        while(m-- != 0) {
            st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());
            int r = Integer.parseInt(st.nextToken());

            i = 1; j = r;
            BigInteger rFact = mul(0, MX, ROOT);
            i = n - r + 1; j = n;
            BigInteger nPr = mul(0, MX, ROOT);
            BigInteger ans = nPr.divide(rFact).mod(MOD);

            bw.write(ans.toString());
            bw.newLine();
        }
        bw.flush();
    }
}